### ISSUE TITLE
artificial neural network (nnet package tuning)

### ISSUE THREAD

**Issue by refree** [refree/parameter-tunning#1]
#전처리 필수{preProcess(train, method = c("knnImpute","center","scale"))}
#종속변수 Factor / 나머지 numeric(Factor가 있다면 dummy화 해야함)
#더미화는 아래 링크 참조
#https://www.analyticsvidhya.com/blog/2016/12/practical-guide-to-implement-machine-learning-with-caret-package-in-r-with-practice-problem/

#튜닝 시작=======================================================
library(caret); library(dplyr); library(randomForest); 
library(xgboost)

cb7 <- read.csv("df.csv" )
summary(cb7$do_violence_in_school)
#- numeric --------#
for(i in 1:length(cb7)) {   cb7[,i] <- as.numeric(cb7[,i]) }
sum(is.na(cb7)); cb7 <- as.data.frame(cb7); 

#pre-processing
preProcPrms <- preProcess(cb7, method = c("center", "scale"))
XY_pre <- predict(preProcPrms, cb7)
XY_pre$do_violence_in_school <- ifelse(XY_pre$do_violence_in_school >= 0, 1, 0)
summary(XY_pre$do_violence_in_school);
str(XY_pre)
train_processed <- XY_pre
#https://www.analyticsvidhya.com/blog/2016/12/practical-guide-to-implement-machine-learning-with-caret-package-in-r-with-practice-problem/
#Converting every categorical variable to numerical using dummy variables
dmy <- dummyVars(" ~ .", data = train_processed,fullRank = T)
train_transformed <- data.frame(predict(dmy, newdata = train_processed))

#Checking the structure of transformed train file
str(train_transformed)
#Converting the dependent variable back to categorical
train_transformed$do_violence_in_school<-as.factor(train_transformed$do_violence_in_school)

#Spliting training set into two parts based on outcome: 75% and 25%
index <- createDataPartition(train_transformed$do_violence_in_school, p=0.75, list=FALSE)
trainSet <- train_transformed[ index,]
testSet <- train_transformed[-index,]

#Feature selection using rfe in caret
control <- rfeControl(functions = rfFuncs,
                      method = "repeatedcv",
                      repeats = 3,
                      verbose = FALSE)
outcomeName<-'do_violence_in_school'
predictors<-names(trainSet)[!names(trainSet) %in% outcomeName]
do_violence_in_school_Profile <- rfe(trainSet[,predictors], trainSet[,outcomeName],
                         rfeControl = control)
do_violence_in_school_Profile

modelLookup(model='nnet')

fitControl <- trainControl(
  method = "repeatedcv",
  number = 10,
  repeats = 3)

#Creating grid
grid <- expand.grid(size=c(1,2,3,4,5,6,7,8,9,10),
                    decay=c(0,0.001,0.1,5e-04)
                    )
#training the model
model_nnet<-train(trainSet[,predictors],trainSet[,outcomeName],method='nnet',metric="Accuracy",
                  trControl=fitControl,tuneGrid=grid)

print(model_nnet); plot(model_nnet)
